go/vt/vtgate/planbuilder/operators/node.go

```diff
@@ -90,3 +91,31 @@ func checkSize(inputs []Operator, shouldBe int) {
 		panic(fmt.Sprintf("BUG: got the wrong number of inputs: got %d, expected %d", len(inputs), shouldBe))
 	}
 }
+
+type rewriterFunc func(*plancontext.PlanningContext, Operator) (newOp Operator, changed bool, err error)
+
+func rewriteBottomUp(ctx *plancontext.PlanningContext, root Operator, rewriter rewriterFunc) (Operator, bool, error) {
+	oldInputs := root.Inputs()
+	anythingChanged := false
+	newInputs := make([]Operator, len(oldInputs))
+	for i, operator := range oldInputs {
+		in, changed, err := rewriteBottomUp(ctx, operator, rewriter)
+		if err != nil {
+			return nil, false, err
+		}
+		if changed {
+			anythingChanged = true
+		}
+		newInputs[i] = in
+	}
+
+	if anythingChanged {
+		root = root.Clone(newInputs)
+	}
+
+	newOp, b, err := rewriter(ctx, root)
+	if err != nil {
+		return nil, false, err
+	}
+	return newOp, anythingChanged || b, nil
+}
```

Comment:
I'm using the variable like this:

```go
	if anythingChanged {
		root = root.Clone(newInputs)
	}
```

Was trying to avoid object creation. Maybe it's not important, didn't actually measure anything

Reply:
Oh, I missed that. Its fine, let's just keep it.